workers/unity/Packages/com.improbable.gdk.core/View/View.cs

```diff
@@ -0,0 +1,102 @@
+        private readonly List<IViewStorage> viewStorages = new List<IViewStorage>();
+
+        private readonly HashSet<EntityId> entities = new HashSet<EntityId>();
+
+        public View()
+        {
+            foreach (var assembly in AppDomain.CurrentDomain.GetAssemblies())
+            {
+                foreach (var type in assembly.GetTypes())
+                {
+                    if (typeof(IViewStorage).IsAssignableFrom(type) && !type.IsAbstract)
+                    {
+                        var instance = (IViewStorage) Activator.CreateInstance(type);
+
+                        typeToViewStorage[instance.GetSnapshotType()] = instance;
+
+                        viewStorages.Add(instance);
+                    }
+                }
+            }
+        }
+
+        public void ApplyDiff(ViewDiff diff)
+        {
+            var entitiesAdded = diff.GetEntitiesAdded();
+            foreach (var entity in entitiesAdded)
+            {
+                entities.Add(entity);
+            }
+
+            var entitiesRemoved = diff.GetEntitiesRemoved();
+            foreach (var entity in entitiesRemoved)
+            {
+                entities.Remove(entity);
+            }
+
+            foreach (var storage in viewStorages)
+            {
+                // Resolve this with an actual diff!
+                storage.ApplyDiff(diff);
+            }
+        }
+
+        public bool HasEntity(EntityId entityId)
+        {
+            return entities.Contains(entityId);
+        }
+
+        // TODO: Make ref readonly when we have a dictionary type with ref indexing semantics.
+        public T GetComponent<T>(EntityId entityId) where T : struct, ISpatialComponentSnapshot
+        {
+            if (!HasEntity(entityId))
+            {
+                throw new ArgumentException($"The view does not have entity with Entity ID: {entityId.Id}");
+            }
+
+            var storage = (IViewComponentStorage<T>) typeToViewStorage[typeof(T)];
+            return storage.GetComponent(entityId.Id);
+        }
+
+        public bool HasComponent<T>(EntityId entityId) where T : struct, ISpatialComponentSnapshot
+        {
+            if (!HasEntity(entityId))
+            {
+                return false;
+            }
+
+            var storage = (IViewComponentStorage<T>) typeToViewStorage[typeof(T)];
+            return storage.HasComponent(entityId.Id);
+        }
+
+        public Authority GetAuthority<T>(EntityId entityId) where T : struct, ISpatialComponentSnapshot
+        {
+            if (!HasEntity(entityId))
+            {
+                throw new ArgumentException($"The view does not have entity with Entity ID: {entityId.Id}");
+            }
+
+            return ((IViewComponentStorage<T>) typeToViewStorage[typeof(T)]).GetAuthority(entityId.Id);
+        }
+
+        public bool IsAuthoritative<T>(EntityId entityId) where T : struct, ISpatialComponentSnapshot
+        {
+            if (!HasComponent<T>(entityId))
```

Comment:
Could just make this GetAuthority<> == Authoritative.
Which also raises if we need this. It doesn't seem bad to have it in now but we might want to make it a hashset of things that have authority or something that can be returned.

Reply:
Hmmm those semantics are slightly different.

Currently:

- `GetAuthority<T>` throws if you don't have the entity/component in your view.
- `IsAuthoritative<T>` just returns false if you don't have the entity/component in your view.

We can unify these by allowing `GetAuthority` to return `NotAuthoritative` if the entity/component isn't in your view and doing as you suggest.

Thoughts?